torch/_functorch/autograd_function.py

```diff
@@ -173,9 +179,58 @@ def mark_dirty_error(*args, **kwargs):
 # the unwrapping to be consistent with regular PyTorch dispatcher operations.
 
 
+class VmapInfo(NamedTuple):
+    batch_size: int
+
+
 @custom_function_call.py_impl(TransformType.Vmap)
 def custom_function_call_vmap(interpreter, autograd_function, *operands):
-    raise RuntimeError("NYI: vmap rule for custom_function_call")
+    if not hasattr(autograd_function, "vmap"):
+        # TODO: link docs when they're ready.
+        # https://github.com/pytorch/pytorch/issues/90224
+        raise RuntimeError(
+            f"You tried to vmap over {autograd_function.__name__}, but "
+            f"it does not have a vmap rule defined. Please add a vmap "
+            f"staticmethod to it.")
+
+    current_level = interpreter.level()
+    info = VmapInfo(batch_size=interpreter.batch_size())
+    unwrapped_operands, in_dims = unwrap_batched(operands, current_level)
+
+    # If none of the tensors are batched at the current level, then we skip the
+    # current level. This saves the user from needing to handle this case in
+    # their vmap staticmethod (and is consistent with our C++ batching rule API)
+    if pytree.tree_all(lambda dim: dim is None, in_dims):
+        with interpreter.lower():
+            return custom_function_call(autograd_function, *operands)
+
+    with interpreter.lower():
+        unwrapped_output, out_dims = autograd_function.vmap(info, in_dims, *unwrapped_operands)
+
+    output = wrap_batched(unwrapped_output, out_dims, current_level)
+    return output
```

Comment:
This is more of a pyfunctorch question, but is there a reason this plumbing cannot be done at the pyfunctorch level? If that is in place, all we need to do here is call `autograd_function.vmap`.

Analogously for autograd, interpreter.lower() would handling the boilerplate of unwrapping and unwrapping.

Reply:
> This is more of a pyfunctorch question, but is there a reason this plumbing cannot be done at the pyfunctorch level? If that is in place, all we need to do here is call autograd_function.vmap.

That's a good question. I can see it going either way.

In C++, there are two types of batching rules that people write:
(1) unwrap(),  lower(), do_something(), wrap()
(2) call a sequence of PyTorch operations

Most of our C++ batching rules are written as the former. CompositeImplicitAutograd and some other things are written as the latter. There are a non-zero amount of rules that are a hybrid between the two (yikes).

pyfunctorch is supposed to be consistent with our C++ dispatcher, so it should in theory let people (developers) do either instead of forcing (1) onto them. For autograd.Function with vmap staticmethod, we've made the decision that we are forcing (1) onto people for simplicity (so they don't need to manually call interpreter.lower() and understand what an interpreter is).

Reply:
Interesting, I didn't think about the case where we want to register a composite. But one workaround could be: if we force (1) on people, maybe we could replace that functionality with an API specifically for registering decompositions for layers. 

Sounds like this might prevent people from doing the hybrid approach, but maybe we want to discourage those anyway, and this would actually be the way of enforcing that?

Reply:
Hybrid approach isn't bad if there's a good reason for it (which there is in the C++ code -- it's difficult to write a batching rule when your operation does broadcasting).

For this PR I'll land it as-is, but we can revisit the situation in the future.

> Analogously for autograd, interpreter.lower() would handling the boilerplate of unwrapping and unwrapping.

To check my understanding of this comment -- would one no longer have an interpreter.lower() inside the custom_function_call_grad function if we gave the same treatment to the grad transform?

Reply:
> To check my understanding of this comment -- would one no longer have an interpreter.lower() inside the custom_function_call_grad function if we gave the same treatment to the grad transform?

Oh what I meant was: In c++ functorch, the "send to next" of autograd interpreter handles unwrapping (which is pop layer off stack + unwrap), but in pyfunctorch the "send to next" for autograd interpreter is just lower (just pop layer of stack), so that is why user has to manually unwrap. I don't know if we want to do anything about that though.